File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* node) {
        // base case
        if(node == NULL) return 0;
        
        int leftHeight = solve(node -> left);
        int rightHeight = solve(node -> right);
        
        return (1 + max(leftHeight, rightHeight));
    }
    int maxDepth(TreeNode* root) {
        
        
        return solve(root);
    }
};

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
    // bfs iterative
    if root == nil {
        return 0
    }
    
    queue := []*TreeNode{root}
    maxDepth := 0
    
    for len(queue) > 0 {
        size := len(queue)

        for i := 0; i < size; i++ {
            curr := queue[0]
            queue = queue[1:]
            
            if curr.Left != nil {
                queue = append(queue, curr.Left)
            }
            
            if curr.Right != nil {
                queue = append(queue, curr.Right)
            }
        }
        
        maxDepth++
    }
    
    return maxDepth
}

// tc -> O(n), n is number of total nodes
// sc -> O(h), h depth of tree